ext/socket/raddrinfo.c

```diff
@@ -752,7 +758,7 @@ rb_getnameinfo(const struct sockaddr *sa, socklen_t salen,
     }
 
     pthread_t th;
-    if (raddrinfo_pthread_create(&th, do_getnameinfo, arg) != 0) {
+    if (raddrinfo_pthread_create(&th, fork_safe_do_getnameinfo, arg) != 0) {
```

Comment:
Do you have the crash info handy? IIRC you said the crash was in `raddrinfo_pthread_create`?

I'm wondering if the lock should be higher or not. For `rb_getaddrinfo` the lock is above the thread spawn and wait.

Reply:
Right, rereading the ticket, the concern in https://github.com/ruby/ruby/pull/10864 /  [[Feature #20590]](https://bugs.ruby-lang.org/issues/20590) was that `getaddrinfo` synchronize a mutex, so the goal was to avoid forking while it is held. And the issue it was trying to avoid is deadlocked child, not a crash in pthread.

I'm OK with moving the lock up so the `pthread` functions are covered, as I don't think it makes a meqaningful difference, but I think we should at least have a theory that would explain this crash.

Reply:
I was thinking, a signal handler is invoked before `pthread_detach()`, it happens to start on the thread running getnameinfo, it also calls `fork`, and the child process (doesn't exit and) moves on to call `pthread_detach()`.

I guess that was an unreasonable guess. But this is all I can come up with for `fork`.

Other than that, my guess is that a thread without GVL or a signal handler somehow discovers and detaches that thread. It doesn't seem like a good guess either, but otherwise `pthread_detach()` wouldn't crash.